### DISCUSSION TITLE
Waterfall optimization

### DISCUSSION THREAD

**Discussion by kof** [remix-run/remix#1914] [Q&A] [answered]
I was looking at Kent's [site](https://kentcdodds.com/) the other day to see how remix performs in some not-so-trivial site/blog. Results are great already, really, super happy this is happening!

One thing that caught my attention is that some waterfall is still happening at the beginning. Before app chunks are loaded, they are waiting for fonts, some styles and polyfill.js costing around 500ms from my machine of waiting time.

Further more I can see despite the Matter font blocking the loading of the rest of the UI, I still see it causing a rerender (if you load in 3g quality for slowmo). 

I am looking forward to the future where we could load a site with a custom font, loaded in a non-blocking way and without layout shifts 🎉

So, I am wondering if remix could have prevented these?

<img width="948" alt="Screenshot 2022-02-11 at 20 19 24" src="https://user-images.githubusercontent.com/52824/153703972-6e245dd2-7a75-417d-b2df-a6a1cd1eeb4e.png">

**Comment by ccssmnn** [comment]
What you see there is no waterfall. It is queuing which is done by the browser. From Chrome's [timing explanation](https://developer.chrome.com/docs/devtools/network/reference/?utm_source=devtools#timing-explanation):

> Queueing. The browser queues requests when:
> - There are higher priority requests.
> - There are already six TCP connections open for this origin, which is the limit. Applies to HTTP/1.0 and HTTP/1.1 only.
> - The browser is briefly allocating space in the disk cache

The white bar is indicating this.

EDIT: We are seeing the network waterfall diagram, but not the negative association of waterfalls, e.g. Network requests being dependent on each other.

**Comment by kof** [comment]
@ccssmnn thank you for the info, super valuable! Didn't think http1 could be the issue ! 

1. I am calling waterfall whatever is blocking the download, like in this case the first js chunk is queued but its not downloading, am I wrong to call this a waterfall? In mean it literally looks like one :)
2. I just enabled the column that shows protocol and chrome tells me it is using http2, now thats even more interesting!

<img width="1558" alt="Screenshot 2022-02-12 at 14 22 58" src="https://user-images.githubusercontent.com/52824/153713182-8d423ad7-2057-4b20-86e3-26d16e795ea4.png">

**Reply by ccssmnn** [reply]
These are all waterfalls. Some are sequential and others are parallel. We want the parallel one. Borrowing from the remix homepage here.

<img width="1298" alt="Screenshot 2022-02-12 at 14 28 51" src="https://user-images.githubusercontent.com/29344072/153713331-3e76b33b-ea95-42e3-8f91-e8537c3bb3db.png">

What we avoid is network requests being dependent on each other intentionally, e.g. having a request fired by client-side JavaScript. This can only start after JavaScript is loaded and parsed. (The left part in the image)

What you see in your screenshots is: the browser could download everything at once but is applying its own optimizations and queuing. Remix is offering everything needed right away. (The right part in the image)

**Comment by kof** [comment]
@ccssmnn Yeah, I was intuitively including only sequencial requests in "waterfall" terminology 🤣

**Comment by kof** [comment]
@ccssmnn in your screenshot from remix landing, everything is parallel, except of document, this is correct. In my screenshots you see Document => (fonts, stylesheets, polyfills ... 7 requests in total) => rest.

I am saying that those 7 requests after the document are blocking the download of the rest. Not sure why you are not seeing it the same way.

**Reply by ccssmnn** [reply]
They are not **blocked**, they are **queued**.

This is done by the browser and has nothing to do with Remix. The browser clearly _could_ download everything at once but is applying **its own prioritization** to the downloads, e.g. ["Higher priority requests"](https://developer.chrome.com/docs/devtools/network/reference/#timing-explanation). Maybe:

- Fonts and stylesheets are higher in prioritization, since they refer to the look of the page. The sooner you loaded these, the faster you can start with the initial paint.
- Polyfills are needed by every other JS, so you should load/parse this first.

Downloading everything at once would slow styles and fonts down and ultimately result in a slower page. These decisions are made by Chrome, not Remix.

**Reply by kof** [reply]
Maybe "blocked" is not the right word, but it definitely has to do with how ressources are described in the document, browser doesn't optimise this magically, its because those 7 are in the header and the chunks are at the end of the body.

That definitely explains why the rerender happens when font is loaded (separate problem though)

I guess this is currently best for low bandwidth, downloading only what's necessary first to avoid using the bandwidth for the scripts that are only needed later

<img width="869" alt="Screenshot 2022-02-12 at 14 58 49" src="https://user-images.githubusercontent.com/52824/153714325-80530bca-4fd6-407c-ad2f-edb1608d7e95.png">

**Reply by ccssmnn** [reply]
Chrome might use `<head>` vs. `<body>` for their queuing prioritization, but I cannot find any documentation on that. Firefox and Safari show different loading orders than Chrome.

**Reply by kof** [reply]
Indeed everything is different in Firefox, it even manages to have the font loaded before rendering text, so there is no rerender for that part.

**Comment by ccssmnn** [comment, marked as answer]
Let me summarize the discussion:

> One thing that caught my attention is that some waterfall is still happening at the beginning. Before app chunks are loaded, they are waiting for fonts, some styles and polyfill.js costing around 500ms from my machine of waiting time.

Resources are not blocked. This is Chrome **queuing** the other resources due to [lower prioritization](https://developer.chrome.com/docs/devtools/network/reference/#timing-explanation), indicated by the white bar. Chrome might prioritize assets in `<head>` over the ones in `<body>` which is why polyfills, fonts and CSS are loaded first. Safari and Firefox behave different because they have different opinions on the best user experience when loading websites.

> Further more I can see despite the Matter font blocking the loading of the rest of the UI, I still see it causing a rerender (if you load in 3g quality for slowmo).

I could not reproduce this. But what you described is a [font-display strategy](https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face/font-display). Kent used [`font-display: swap`](https://github.com/kentcdodds/kentcdodds.com/blob/579fd40a7c4a3852ffd1b61f7d3a737bdcf2b6d1/styles/tailwind.css#L16), which shows the system default font until the font is loaded. Chrome has waited for the CSS to load to show a styled UI on the first render, but the font needed a little longer to load and was replaced as soon as the font arrived.

> I am looking forward to the future where we could load a site with a custom font, loaded in a non-blocking way and without layout shifts.

Choose a different font display strategy.

> So, I am wondering if remix could have prevented these?

Chromes resource prioritization and queuing? No. Showing the system font instead of the custom font until it is loaded? Possible with standard CSS.

**Reply by kof** [reply]
The best strategy for fonts is probably new [size-adjust property](https://www.smashingmagazine.com/2021/05/reduce-font-loading-impact-css-descriptors/) which lets us have a system font adjusted to the metrics of the custom font and when its loaded the swap won't cause recalc, currently that's exactly what is happening because of swap.

**Reply by kof** [reply]
Regarding ressource prioritization - yes and no. We can communicate different strategies to the browser, the question is if we can make any better choices than what's on the page right now (idk). If I were optimizing for a high bandwidth user, I would not have waited 500ms at the beginning and would have started all of the ressources download in parallel. This would have reduced the TTI (js based) by 500ms for this site, but yeah, its not necessarily needed if the site is architected in a way that those interactive chunks in js are optional and I could potentially interract without them, so its sort of interractive already from the the beginning in a pure html way.

**Reply by ccssmnn** [reply]
The parallel vs queued downloads are choices made by the browser, not the web developer. Prioritizing font and styles for a fast FCP sounds perfectly reasonable to me. Adding polyfills to support older browsers [here](https://github.com/kentcdodds/kentcdodds.com/blob/579fd40a7c4a3852ffd1b61f7d3a737bdcf2b6d1/app/root.tsx#L335) is a decision based on what you want to support. Polyfills must be parsed BEFORE your other scripts to achieve the desired effect, which is why he is including them in the head.

You are free to make all the decisions with Remix (e.g. putting `<Scripts/>` in the `<head>`), what I really like. What we see is nothing caused by Remix, and therefore nothing Remix could have prevented. Just browsers being browsers, and CSS behaving as intended.

**Reply by kof** [reply]
Remix can have a higher level abstraction that decides where to put which script and which css file (well for styles it already has `links`). Browsers did the choices in that case but only based on how things are structured in the document. 

e.g. we can communicate to the browser to start downloading the polyfills in parallel with other scripts but evaluate them in the right order (defer)

**Reply by ccssmnn** [reply]
Remix [links](https://remix.run/docs/en/v1/api/conventions#links) are not an abstraction, and not an abstraction for styles. They are a 1:1 interface to the standard [HTML `<link>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/link) element for external resources. This is needed because you need to be able to set `<link>` elements in your routes without having access to `<head>`, which is in `root`. Just like the other methods in the [Route Module API](https://remix.run/docs/en/v1/api/conventions#route-module-api).

The whole idea about Remix is not being an abstraction layer over web standards. Instead, embracing them.

**Reply by kof** [reply]
This is not entirely true, but I will leave it by that ;)

**Reply by ccssmnn** [reply]
[Remix: Don't over abstract](https://remix.run/docs/en/v1/pages/philosophy#dont-over-abstract)
